### PR TITLE
Support EngineClient redirection

### PR DESCRIPTION
**Pull request by Joannis**

### CODE REVIEW

File: Sources/Vapor/Engine/EngineClient.swift [hunk 1]
```diff
@@ -24,17 +24,52 @@ public final class EngineClient: Client, Service {
...
+            }
+        }
+    }
+    
+    private func response(
+        from httpRes: HTTPResponse,
+        for req: Request,
+        redirecting: Int
+    ) throws -> Future<Response> {
+        if httpRes.status.code >= 301 && httpRes.status.code <= 303 {
```
Comment (gwynne) [inline Sources/Vapor/Engine/EngineClient.swift]: This does not cover the range of potential redirect responses. Most of the 3xx range is relevant, particularly 307 and 308.

File: Sources/Vapor/Engine/EngineClient.swift [hunk 2]
```diff
@@ -24,17 +24,52 @@ public final class EngineClient: Client, Service {
...
+            }
+            
+            guard let location = httpRes.headers[.location] else {
+                throw VaporError(
+                    identifier: "invalid-redirect",
+                    reason: "The HTTP Client received a status 3xx without a location to redirect to."
+                )
             }
+            
+            req.http.uri = try location.makeURI()
```
Comment (gwynne) [inline Sources/Vapor/Engine/EngineClient.swift]: It is incorrect but still-common practice for servers to return "relative" redirects, e.g. `Location: /path/on/same/server`. This case should be detected and the URI defaulted to the previous URI's schema and hostname.

File: Sources/Vapor/Engine/EngineClient.swift [hunk 3]
```diff
@@ -86,10 +119,16 @@ public struct EngineClientConfig: Service {
...
+    
+    /// The maximum amount of 3xx redirect responses to follow
+    ///
+    /// Used to prevent infinite redirect loops
+    public var maxRedirections: Int
 
     /// Create a new EngineClientConfig.
     public init(maxResponseSize: Int) {
         self.maxResponseSize = maxResponseSize
+        self.maxRedirections = 3
```
Comment (gwynne) [inline Sources/Vapor/Engine/EngineClient.swift]: I would suggest a slightly higher default, probably 8

File: Tests/VaporTests/ApplicationTests.swift [hunk 4]
```diff
@@ -67,10 +67,20 @@ class ApplicationTests: XCTestCase {
...
     }
+    
+    func testClientRedirect() throws {
+        let app = try Application()
+        
+        let client = try app.make(Client.self)
+        
+        let response = try client.get("http://www.google.com/").blockingAwait()
+        XCTAssertEqual(response.http.status, 200)
+    }
```
Comment (gwynne) [inline Tests/VaporTests/ApplicationTests.swift]: Add tests for the "relative" redirect case and for the various 3xx codes.

File: Sources/Vapor/Engine/EngineClient.swift [hunk 5]
```diff
@@ -24,17 +29,88 @@ public final class EngineClient: Client, Service {
...
+        from httpRes: HTTPResponse,
+        for req: Request,
+        redirecting: Int
+    ) throws -> Future<Response> {
+        if httpRes.status.code >= 300 && httpRes.status.code < 400 {
+            switch httpRes.status.code {
+            case 301, 307, 308:
+                return try redirect(httpRes, for: req, redirecting: redirecting)
+            case 302, 303:
+                req.http.method = .get
```
Comment (gwynne) [inline Sources/Vapor/Engine/EngineClient.swift]: When transforming to GET, make sure to remove any request body and Content-Type header.

**Comment by Joannis** [comment]
` [TCP.TCPError.getAddressInfo: Success] ` is a really great error

File: Sources/Vapor/Engine/EngineClient.swift [hunk 6]
```diff
@@ -51,16 +129,15 @@ public final class EngineClient: Client, Service {
...
-            res.http = httpRes
-            return res
+            
+            return try self.response(from: httpRes, for: req, redirecting: redirecting)
         }
     }
 
     /// Responds to a Request using TCP client.
-    private func plaintextRespond(to req: Request) throws -> Future<Response> {
+    private func plaintextRespond(to req: Request, redirecting: Int) throws -> Future<Response> {
```
Comment (tanner0101) [inline Sources/Vapor/Engine/EngineClient.swift]: I know this is internal API, but I think it should be `maxRedirections:`

File: Sources/Vapor/Engine/EngineClient.swift [hunk 7]
```diff
@@ -24,17 +29,90 @@ public final class EngineClient: Client, Service {
...
+        guard redirecting > 0 else {
+            throw VaporError(
+                identifier: "excessive-redirects",
+                reason: "The HTTP Client was redirected more than \(config.maxRedirections) times."
+            )
+        }
+        
+        guard let location = response.headers[.location] else {
+            throw VaporError(
+                identifier: "invalid-redirect",
```
Comment (tanner0101) [inline Sources/Vapor/Engine/EngineClient.swift]: camelCase 😭

File: Sources/Vapor/Engine/EngineClient.swift [hunk 8]
```diff
@@ -24,17 +29,90 @@ public final class EngineClient: Client, Service {
...
+        let newURI = try location.makeURI()
+        
+        if newURI.hostname != nil {
+            req.http.uri = newURI
+        } else {
+            if newURI.path.first == "/" {
+                req.http.uri.path = newURI.path
             } else {
-                return try self.plaintextRespond(to: req)
+                var path = newURI.path
```
Comment (tanner0101) [inline Sources/Vapor/Engine/EngineClient.swift]: can this chunk of code get moved to an extension/method that explains what it does?

Comment (tanner0101) [reply Sources/Vapor/Engine/EngineClient.swift]: or a comment..

Comment (gwynne) [reply Sources/Vapor/Engine/EngineClient.swift]: I would go with a well-named extension method myself, yeah.

File: Sources/Vapor/Engine/EngineClient.swift (hunk 6 above)
Comment (gwynne) [reply Sources/Vapor/Engine/EngineClient.swift]: Agreed, `redirecting` reads more like a boolean value than a limit count.

**Comment by tanner0101** [comment]
Also needs tests passing

**Comment by Joannis** [comment]
Ping @tanner0101 it was working by the way, but I guess something changed causing the `Application` eventloop to be idle by default. I changed the tests accordingly.

**Pull request merged**